### DISCUSSION TITLE
Error can't be exported

### DISCUSSION THREAD

**Discussion by elrondwong** [ceph/go-ceph#557] [Ideas]
Error can't export, type assert will not be use.

**Comment by phlogistonjohn** [comment]
Could you please provide more detail? I'm afraid your message is too vague for us to interpret. If possible provide code examples of what you're trying to do. Thanks!

**Reply by elrondwong** [reply]
All errors can't be exported
- rgw/errors.go
- rados/errors.go
- ...

example:
call  rgw admin.RemoveUser  -> admin.call  

```go
// admin.call
	if resp.StatusCode >= 300 {
		return nil, handleStatusError(decodedResponse)
	}
```

admin.call return statusError

When I want to use assertions to implement error classification processing, I find that statusError cannot be exported.
If statusError can be exported. error handle code:

```go
if err != nil {
    if rgwErr, ok := err.(statusError); ok {
        switch statusError.Code {
 		case "InvalidAccessKeyId":
			log.Logger.Warn("user does not exist")
                case "xxxx":
                        xxxx
                        ...
		default:
			log.Logger.Error(err.Error())
			return err
        }
    }
}
```

But now I have to re-parse：

```go
type UserError struct {
	Code      string `json:"Code"`
	HostID    string `json:"HostID"`
	RequestID string `json:"RequestID"`
}

if err != nil {
		uErr := UserError{}
		bytes, _ := json.Marshal(err)
		_ = json.Unmarshal(bytes, &uErr)
    if rgwErr, ok := err.(uErr); ok {
        switch uErr.Code {
 		case "InvalidAccessKeyId":
			log.Logger.Warn("user does not exist")
                case "xxxx":
                        xxxx
                        ...
		default:
			log.Logger.Error(err.Error())
			return err
        }
    }
}

```

If it can be exported like the aws s3 sdk error, it can be much simpler-

Thank you!

**Reply by phlogistonjohn** [reply]
Thank you. First, rados errors and rgw admin errors are a bit different. I'll start with rados errors since I'm more familiar with that:

Not exporting the rados error type is a deliberate choice. For certain common errors we have provided "sentinel values" like `ErrNotFound`. However, because we can't handle every error condition we do export a public method `ErrorCode` that can be used like so:
```go
type errorWithCode interface {
    ErrorCode() int
}
err := rados.SomeRadosFunc()
var ec errorWithCode
if errors.As(err, &ec) {
   errCode := ec.ErrorCode()
   // ... do something with errCode ...
}
```

Now, as for the rgw admin errors, that library doesn't work the same way as it always goes over http(s) and thus the approach to errors is different.

To translate your example into something that should work with go-ceph's rgw admin module today. I think the following ought to work (but I have not tested it myself):

```go
err := rgw.SomeAPICall()
if err != nil {
    if errors.Is(err, rgw.ErrInvalidAccessKey) {
       // ... do something specific to access errors ...
    } else if errors.Is(err, rgw.ErrNoSuchUser) {
       // ... do something specific to user not existing ...
    } else {
       // ... do something general ...
    }
}
```

If you also need RequestID and/or HostID values, I could see an issue, but just telling the error conditions apart should be doable with the library as-is.

In the meantime I'll highlight @leseb and @thotz who are our subject matter experts on the new rgw admin package. We'll see if they've got anything additional to add, but I do encourage you to try the approach using `errors.Is` and see if that works for you.

Good luck!

**Reply by elrondwong** [reply]
Thank you very much!
I‘m very glad to acquire knowledge from above！
Have a nice weekend!

**Reply by thotz** [reply]
I totally agree with @phlogistonjohn and u can find such examples in [rook repo](https://github.com/rook/rook/tree/master/pkg/operator/ceph/object)
